pilot/pkg/config/kube/ior/route.go

```diff
@@ -627,3 +331,143 @@ func hostHash(name string) string {
 	hash := sha256.Sum256([]byte(name))
 	return hex.EncodeToString(hash[:8])
 }
+
+func (r *routeController) reconcileGateway(config *config.Config, routes []*v1.Route) error {
+	gateway, ok := config.Spec.(*networking.Gateway)
+
+	if !ok {
+		return fmt.Errorf("could not decode spec as Gateway from %v", config)
+	}
+
+	var serviceNamespace string
+	var serviceName string
+	var err error
+
+	serviceNamespace, serviceName, err = r.findService(gateway)
+
+	if err != nil {
+		return errors.Wrapf(err, "gateway %s/%s does not specify a valid service", config.Namespace, config.Name)
+	}
+
+	routeMap := make(map[string]*v1.Route)
+
+	for _, v := range routes {
+		routeMap[v.Name] = v
+	}
+
+	var result *multierror.Error
+
+	for _, server := range gateway.Servers {
+		for _, host := range server.Hosts {
+			var err error
+
+			name := getRouteName(config.Namespace, config.Name, host)
+
+			_, found := routeMap[name]
+
+			if found {
+				_, err = r.updateRoute(config.Meta, host, server.Tls, serviceNamespace, serviceName)
```

Comment:
You need to pass the route from `routeMap` to `updateRoute` so that you can properly update it. To update an object, a controller must preserve the metadata of the existing object (this is especially true for the `resourceVersion`, but most other fields should be preserved as well.